lib/model/folder_sendrecv.go

```diff
@@ -1334,16 +1294,85 @@ func (f *sendReceiveFolder) copierRoutine(in <-chan copyBlocksState, pullChan ch
 				state.copyDone(block)
 			}
 		}
-		if file != nil {
-			// os.File used to return invalid argument if nil.
-			// fs.File panics as it's an interface.
-			file.Close()
-		}
 
 		out <- state.sharedPullerState
 	}
 }
 
+// copyByWeakHash copies all blocks that can be found from their weak hash.
+// It returns an updated state with these blocks removed.
+func (f *sendReceiveFolder) copyByWeakHash(state copyBlocksState, w io.WriterAt) (copyBlocksState, error) {
+	file, err := f.fs.Open(state.file.Name)
+	if err != nil {
+		return state, err
+	}
+	defer file.Close()
+
+	buf := protocol.BufferPool.Get(state.file.BlockSize())
+	defer protocol.BufferPool.Put(buf)
+
+	finder := weakhash.NewFinder(file, buf)
+
+	byWeakhash := make(map[uint32][]int) // Block indices by weak hash.
+	for i, block := range state.blocks {
+		if h := block.WeakHash; h != 0 {
+			finder.Add(h)
+			byWeakhash[h] = append(byWeakhash[h], i)
+		}
+	}
+
+	if len(byWeakhash) == 0 {
+		l.Debugf("not weak hashing %s. file did not contain any weak hashes", state.file.Name)
+		return state, nil
+	}
+
+	done := make([]bool, len(state.blocks))
+
+	for finder.Next() {
+		h, offset := finder.Match()
+		sha := sha256.Sum256(buf)
+
+		for _, i := range byWeakhash[h] {
+			if !bytes.Equal(state.blocks[i].Hash, sha[:]) {
+				continue
+			}
+
+			_, err = f.limitedWriteAt(w, buf, offset)
+			if err != nil {
+				state.fail(errors.Wrap(err, "dst write"))
+				return state, err
+			}
+
+			if offset == state.blocks[i].Offset {
+				state.copiedFromOrigin()
+			} else {
+				state.copiedFromOriginShifted()
+			}
+			done[i] = true
+			break
+		}
+
+		select {
+		case <-f.ctx.Done():
+			return state, f.ctx.Err()
+		default:
+		}
+	}
+
+	// Remove the blocks we've done from blocks.
+	blocksleft := state.blocks[:0]
+	for _, indices := range byWeakhash {
+		for _, i := range indices {
+			if !done[i] {
+				blocksleft = append(blocksleft, state.blocks[i])
+			}
+		}
+	}
+	state.blocks = blocksleft
```

Comment:
I think this screws up the remaining block order?

I think you could just iterate over done?